drivers/clock_control/nrf_power_clock.c

```diff
@@ -1,280 +1,327 @@
 /*
- * Copyright (c) 2016 Nordic Semiconductor ASA
+ * Copyright (c) 2016-2019 Nordic Semiconductor ASA
  * Copyright (c) 2016 Vinayak Kariappa Chettimada
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
 #include <soc.h>
-#include <errno.h>
-#include <sys/atomic.h>
-#include <device.h>
 #include <drivers/clock_control.h>
-#include <sys/__assert.h>
-#include <nrf_clock.h>
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
-#include <nrf_power.h>
 #include <drivers/clock_control/nrf_clock_control.h>
-#endif
+#include <logging/log.h>
+#include <nrf_power.h>
 
-static u8_t m16src_ref;
-static u8_t m16src_grd;
-static u8_t k32src_initialized;
+LOG_MODULE_REGISTER(clock_control, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+/* Helper logging macros which prepends device name to the log. */
+#define CLOCK_LOG(lvl, dev, ...) \
+	LOG_##lvl("%s: " GET_ARG1(__VA_ARGS__), dev->config->name \
+			COND_CODE_0(NUM_VA_ARGS_LESS_1(__VA_ARGS__),\
+					(), (, GET_ARGS_LESS_1(__VA_ARGS__))))
+#define ERR(dev, ...) CLOCK_LOG(ERR, dev, __VA_ARGS__)
+#define WRN(dev, ...) CLOCK_LOG(WRN, dev, __VA_ARGS__)
+#define INF(dev, ...) CLOCK_LOG(INF, dev, __VA_ARGS__)
+#define DBG(dev, ...) CLOCK_LOG(DBG, dev, __VA_ARGS__)
+
+/* returns true if clock stopping or starting can be performed. If false then
+ * start/stop will be deferred and performed later on by handler owner.*/
+typedef bool (*nrf_clock_stop_handler_t)(struct device *dev);
```

Comment:
The name of this type is no longer adequate, now that it is used also for checking if starting can be performed.

Reply:
renamed to `nrf_clock_handler_t`